rfcs/0050-vm-syscalls-3/0050-vm-syscalls-3.md

```diff
@@ -0,0 +1,214 @@
+    - `0x0100000000000001`: input cells with the same running script as current script
+    - 2: output cells.
+    - `0x0100000000000002`: output cells with the same running script as current script
+    - 3: dep cells.
+- `bounds`: high 32 bits means `offset`, low 32 bits means `length`. If `length` equals to zero, it read to end instead of reading 0 bytes.
+- `argc`: argc contains the number of arguments passed to the program
+- `argv`: argv is a one-dimensional array of strings
+- `memory_limit`: an integer value denoting the memory size to use(Not including descendant children scripts), possible values include:
+    - 1 (0.5 M)
+    - 2 (1 M)
+    - 3 (1.5 M)
+    - 4 (2 M)
+    - 5 (2.5 M)
+    - 6 (3 M)
+    - 7 (3.5 M)
+    - 8 (4 M)
+- `exit_code`: an int8 pointer denoting where we save the exit code of a child script.
+- `content`: a pointer to a buffer in VM memory space denoting where we would load the sub-script data. The child script will write data in this buffer via `set_content`.
+- `content_length`: a pointer to a 64-bit unsigned integer in VM memory space. When calling the syscall, this memory location should store the length of the buffer specified by `content`. When returning from the syscall, CKB VM would fill in `content_length` with the actual length of the buffer. `content_length` up to 256K.
+
+The arguments used here `index`, `source`, `bounds`, `argc` and `argv` follow the usage described in [EXEC].
+
+This syscall might return the following results:
+
+- 0: Success.
+- 1-3: Reserved. These values are already assigned to other syscalls.
+- 4: Elf format error
+- 5: Exceeded max content length.
+- 6: Wrong memory limit
+- 7: Exceeded max peak memory
+
+Note that now we have a new limit called *Peak Memory Usage*. The maximum memory usage of the parent script and its descendant children cannot exceed this value. Currently, this limit is set at 32M.
+
+Unlike cycles which always increase, the current memory can decrease or increase. When a child script is returned, the occupied memory is freed. This makes current memory usage lower.
+
+
+### Get Memory Limit
+[Get Memory Limit]: #get-memory-limit
+
+Get the maximum available memory for the current script.
+
+```c
+int ckb_get_memory_limit();
+```
+
+For the script(prime script) directly invoked by CKB, it will always return 8(4M). For the child script invoked by prime script or other child script, it depends on the parameters set by *Spawn*.
+
+### Current Memory
+[Current Memory]: #current-memory
+
+Get the Current Memory Usage. The result is the sum of the memory usage of the parent script and the child script.
+
+```c
+int ckb_current_memory();
+```
+
+The system call returns an integer. Multiply it by 0.5 to get the actual memory usage (unit: megabytes)
+
+### Set Content
+[Set Content]: #set-content
+
+The child script can return bytes data to the parent script through `Set Content`.
+
+```c
+int ckb_set_content(uint8_t* content, uint64_t* length);
```

Comment:
What's the return result of this syscall? What happens when it is called in the root script?

Reply:
Leave a reference to the document on `memory_limit`

Reply:
It returns 0(success) or others. Nothing happens when calling set_content in the root script.